wait/host_port_test.go

```diff
@@ -456,55 +463,166 @@ func TestHostPortStrategyFailsWhileInternalCheckingDueToUnexpectedContainerStatu
+					"80": []nat.PortBinding{
+						{
+							HostIP:   "0.0.0.0",
+							HostPort: port.Port(),
 						},
 					},
 				},
-			}, nil
-		},
-		MappedPortImpl: func(_ context.Context, _ nat.Port) (nat.Port, error) {
-			return port, nil
-		},
-		StateImpl: func(_ context.Context) (*types.ContainerState, error) {
-			return &types.ContainerState{
-				Running: true,
-			}, nil
+			},
 		},
-		ExecImpl: func(_ context.Context, _ []string, _ ...exec.ProcessOption) (int, io.Reader, error) {
-			// This is the error that would be returned if the shell is not installed.
-			return 126, nil, nil
+	}, nil)
+	target.EXPECT().MappedPort(anyContext, nat.Port("80")).Return(port, nil)
+	target.EXPECT().State(anyContext).Return(&types.ContainerState{
+		Running: true,
+	}, nil)
+	target.EXPECT().Exec(anyContext, []string{"/bin/sh", "-c", "true && (\n\t\t\t\t\tcat /proc/net/tcp* | awk '{print $2}' | grep -i :0050 ||\n\t\t\t\t\tnc -vz -w 1 localhost 80 ||\n\t\t\t\t\t/bin/sh -c '</dev/tcp/localhost/80'\n\t\t\t\t)\n\t\t\t\t"}).Return(wait.ExitEaccess, nil, nil)
+	target.EXPECT().CopyFileFromContainer(anyContext, testFilename).Return(nil, errNotFound)
+
+	wg := wait.NewHostPortStrategy("80").
+		WithStartupTimeout(5 * time.Second).
+		WithPollInterval(100 * time.Millisecond)
+
+	err = wg.WaitUntilReady(context.Background(), target)
+	require.NoError(t, err)
+}
+
+func TestHostPortStrategySucceedsGivenShellIsNotFound(t *testing.T) {
+	listener, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+	defer listener.Close()
+
+	rawPort := listener.Addr().(*net.TCPAddr).Port
+	port, err := nat.NewPort("tcp", strconv.Itoa(rawPort))
+	require.NoError(t, err)
+
+	target := &mockStrategyTarget{}
+	target.EXPECT().Host(anyContext).Return("localhost", nil)
+	target.EXPECT().Inspect(anyContext).Return(&types.ContainerJSON{
+		NetworkSettings: &types.NetworkSettings{
+			NetworkSettingsBase: types.NetworkSettingsBase{
+				Ports: nat.PortMap{
+					"80": []nat.PortBinding{
+						{
+							HostIP:   "0.0.0.0",
+							HostPort: port.Port(),
+						},
+					},
+				},
+			},
 		},
-	}
+	}, nil)
+	target.EXPECT().MappedPort(anyContext, nat.Port("80")).Return(port, nil)
+	target.EXPECT().State(anyContext).Return(&types.ContainerState{
+		Running: true,
+	}, nil)
+	target.EXPECT().Exec(anyContext, []string{"/bin/sh", "-c", "true && (\n\t\t\t\t\tcat /proc/net/tcp* | awk '{print $2}' | grep -i :0050 ||\n\t\t\t\t\tnc -vz -w 1 localhost 80 ||\n\t\t\t\t\t/bin/sh -c '</dev/tcp/localhost/80'\n\t\t\t\t)\n\t\t\t\t"}).Return(wait.ExitCmdNotFound, nil, nil)
+	target.EXPECT().CopyFileFromContainer(anyContext, testFilename).Return(nil, errNotFound)
 
-	wg := NewHostPortStrategy("80").
+	wg := wait.NewHostPortStrategy("80").
 		WithStartupTimeout(5 * time.Second).
 		WithPollInterval(100 * time.Millisecond)
 
 	err = wg.WaitUntilReady(context.Background(), target)
 	require.NoError(t, err)
 }
+
+func TestInternalCheckFailsGivenShellIsNotInstalled(t *testing.T) {
```

Comment:
These are the exception to the rule 🙃sorry to mess you about. If you undo the test package change they will all just work without making the internal check public

Reply:
This could be testing using the end to end instead of direct call to the method.

Reply:
I've removed these tests and merged them with the existing ones.